datafusion/physical-plan/src/sorts/partial_sort.rs

```diff
@@ -0,0 +1,1004 @@
+    /// Fetch top N results
+    fetch: Option<usize>,
+    /// Whether the stream has finished returning all of its data or not
+    is_closed: bool,
+    /// Execution metrics
+    baseline_metrics: BaselineMetrics,
+}
+
+impl Stream for PartialSortStream {
+    type Item = Result<RecordBatch>;
+
+    fn poll_next(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Self::Item>> {
+        let poll = self.poll_next_inner(cx);
+        self.baseline_metrics.record_poll(poll)
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        // we can't predict the size of incoming batches so re-use the size hint from the input
+        self.input.size_hint()
+    }
+}
+
+impl RecordBatchStream for PartialSortStream {
+    fn schema(&self) -> SchemaRef {
+        self.input.schema()
+    }
+}
+
+impl PartialSortStream {
+    fn poll_next_inner(
+        self: &mut Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Result<RecordBatch>>> {
+        if self.is_closed {
+            assert_eq!(0, self.input_batch.num_rows());
+            return Poll::Ready(None);
+        }
+        let result = match ready!(self.input.poll_next_unpin(cx)) {
+            Some(Ok(batch)) => {
+                self.input_batch = compute::concat_batches(
+                    &self.schema(),
+                    &[self.input_batch.clone(), batch],
+                )?;
+                let slice_point =
+                    self.get_slice_point(self.common_prefix_length, &self.input_batch)?;
+                if slice_point == self.input_batch.num_rows() {
+                    Ok(RecordBatch::new_empty(self.schema()))
+                } else {
+                    self.emit(slice_point)
+                }
+            }
+            Some(Err(e)) => Err(e),
+            None => {
+                self.is_closed = true;
+                // once input is consumed, sort the rest of the inserted batches in input_batch
+                self.emit(self.input_batch.num_rows())
+            }
+        };
+
+        Poll::Ready(Some(result))
+    }
+
+    fn emit(self: &mut Pin<&mut Self>, slice_point: usize) -> Result<RecordBatch> {
+        let original_num_rows = self.input_batch.num_rows();
+        let sliced = self.input_batch.slice(0, slice_point);
+        self.input_batch = self
+            .input_batch
+            .slice(slice_point, self.input_batch.num_rows() - slice_point);
+        let sort_columns = self
+            .expr
+            .clone()
+            .iter()
+            .map(|expr| expr.clone().evaluate_to_sort_column(&sliced))
+            .collect::<Result<Vec<_>>>()?;
+        let sorted_indices = lexsort_to_indices(&sort_columns, None)?;
+        let sorted_columns = sliced
+            .columns()
+            .iter()
+            .map(|c| take(c.as_ref(), &sorted_indices, None))
+            .collect::<Result<_, _>>()?;
+        let result = RecordBatch::try_new(self.schema(), sorted_columns)?;
```

Comment:
For this section, I think you can use `pub(crate) fn sort_batch` util  under `sort.rs` file.